plugin/output/s3/s3.go

```diff
@@ -0,0 +1,269 @@
+	Bucket    string `json:"bucket" required:"true"`
+	Secure    bool   `json:"secure" default:"false"`
+
+	//for mock client injection
+	client *objectStoreClient
+}
+
+func init() {
+	fd.DefaultPluginRegistry.RegisterOutput(&pipeline.PluginStaticInfo{
+		Type:    "s3",
+		Factory: Factory,
+	})
+}
+
+func Factory() (pipeline.AnyPlugin, pipeline.AnyConfig) {
+	return &Plugin{}, &Config{}
+}
+
+func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.OutputPluginParams) {
+	p.controller = params.Controller
+	p.logger = params.Logger
+	p.config = config.(*Config)
+
+	//set up compression
+	p.config.CompressionType = strings.ToLower(p.config.CompressionType)
+	getNewCompressor, ok := compressors[p.config.CompressionType]
+	if !ok {
+		p.logger.Panicf("compression type: %s is not supported", p.config.CompressionType)
+	}
+	p.compressor = getNewCompressor(p.logger)
+
+	dir, f := filepath.Split(p.config.FileConfig.TargetFile)
+
+	p.targetDir = dir
+	p.fileExtension = filepath.Ext(f)
+	p.fileName = f[0 : len(f)-len(p.fileExtension)]
+
+	p.uploadCh = make(chan string, p.config.FileConfig.WorkersCount_*4)
+	p.compressCh = make(chan string, p.config.FileConfig.WorkersCount_)
+
+	for i := 0; i < p.config.FileConfig.WorkersCount_; i++ {
+		go p.uploadWork()
+		go p.compressWork()
+	}
+
+	//Initialize minio client object.
+	minioClient, err := minio.New(p.config.Endpoint, p.config.AccessKey, p.config.SecretKey, p.config.Secure)
+	p.client = minioClient
+	if err != nil || minioClient == nil {
+		p.logger.Panicf("could not create minio client, error: %s", err.Error())
+	}
+
+	if p.config.client != nil {
+		p.logger.Info("set mock client")
+		p.client = *p.config.client
+	}
+
+	exist, err := p.client.BucketExists(p.config.Bucket)
+	if err != nil {
+		p.logger.Panicf("could not check bucket: %s, error: %s", p.config.Bucket, err.Error())
+	}
+	if !exist {
+		p.logger.Panicf("bucket: %s, does not exist", p.config.Bucket)
+	}
+	p.logger.Info("client is ready")
+	p.logger.Infof("bucket: %s exists", p.config.Bucket)
+
+	anyPlugin, _ := file.Factory()
+	p.outPlugin = anyPlugin.(*file.Plugin)
+
+	p.outPlugin.SealupCallback = p.addFileJob
+
+	p.outPlugin.Start(&p.config.FileConfig, params)
+	p.uploadExistedFiles()
+}
+
+func (p *Plugin) Stop() {
+	p.outPlugin.Stop()
+}
+
+func (p *Plugin) Out(event *pipeline.Event) {
+	p.outPlugin.Out(event)
+}
+
+// uploadFiles gets files from dirs, sorts it, compresses it if it's need, and then upload to s3
+func (p *Plugin) uploadExistedFiles() {
+	//compress all files that we have in the dir
+	p.compressFilesInDir()
+	// get all compressed files
+	pattern := fmt.Sprintf("%s*%s", p.targetDir, p.compressor.getExtension())
+	compressedFiles, err := filepath.Glob(pattern)
+	if err != nil {
+		p.logger.Panicf("could not read dir: %s", p.targetDir)
+	}
+	//sort compressed files by creation time
+	sort.Slice(compressedFiles, p.getSortFunc(compressedFiles))
+	//upload archive
+	for _, z := range compressedFiles {
+		p.uploadCh <- z
```

Comment:
send files further to upload in the compressing stage
if you do it here you'll wait till all files will be compressed and only after that you call uploading stage